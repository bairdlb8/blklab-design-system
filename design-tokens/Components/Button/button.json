{
  "ds": {
    "button": {
      "padding": {
        "horizontal": {
          "value": "16px"
        }
      },
      "color": {
        "primary": {
          "foreground": {
            "value": "{ds.color.text.base.value}"
          },
          "background": {
            "value": "{ds.color.primary.value}"
          }
        },
        "secondary": {
          "foreground": {
            "value": "{ds.color.text.white.value}"
          },
          "background": {
            "value": "{ds.color.secondary.value}"
          }
        }
      }
    }
  }
}
