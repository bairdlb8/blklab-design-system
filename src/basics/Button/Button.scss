.ds-button {
  background-color: #000;
  color: #fff;
  padding: var(--spacer-2) var(--spacer-4);
  border: solid 1px transparent;
  box-shadow: var(--shadow);
  transition: all 0.25s linear;
  line-height: 1;
  position: relative;

  .root {
    width: 100%;
    display: inline-block;
  }

  .root > span {
    display: inline-flex;
    align-items: center;
  }

  &[ds-icon-side='left'] {
    .ds-icon {
      margin-right: 1.25em;
    }
  }

  &[ds-icon-side='right'] {
    .ds-icon {
      margin-left: 1.25em;
    }
  }

  &:disabled {
    background-color: var(--gray-color-lighter);
    color: var(--gray-color-dark);
    cursor: not-allowed;
  }

  &[ds-primary='true'] {
    background-color: var(--primary-color);
    color: var(--button-text-color-light, #fff);
    &:hover {
      background-color: var(--primary-color-dark);
    }

    &:focus {
    }
  }

  &[ds-secondary='true'] {
    background-color: var(--secondary-color);
    color: var(--primary-color-dark);
    &:hover {
      background-color: var(--secondary-color-dark);
    }

    &:focus {
    }
  }

  &.loading {
  }

  &.destructive {
    &:hover {
    }

    &:focus {
    }
  }

  &[ds-hollow='true'] {
    background-color: transparent;
    color: var(--text-color-default);
    border-color: var(--text-color-default);
    &:hover {
      background-color: var(--primary-color);
    }

    &:focus {
    }
  }

  &.display-block {
    width: 100%;
  }

  &.submit {
  }

  &[ds-size='small'] {
    font-size: var(--body-size);
    padding: var(--spacer-2) var(--spacer-3);
  }

  &[ds-size='medium'] {
    font-size: var(--body-size);
    padding: var(--spacer-3) var(--spacer-4);
    font-size: 1.25em;
    svg {
      height: 1.15em;
      width: 1.15em;
    }
  }

  &[ds-size='large'] {
    font-size: var(--body-size);
    padding: var(--spacer-3) var(--spacer-5);
    font-size: 1.35em;
    svg {
      height: 1.25em;
      width: 1.25em;
    }
  }

  &.icon {
    &.left {
    }

    &.right {
    }
  }
}
