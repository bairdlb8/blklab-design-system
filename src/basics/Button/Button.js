/**
- Test thing
- Another test.
**/

import React, { useState } from 'react'
import PropTypes from 'prop-types'
import './Button.scss'

import Icon from '../Icon'

const IconSides = {
  left: 'left',
  right: 'right'
}

const Alignments = {
  left: 'left',
  right: 'right',
  center: 'center'
}

const Sizes = {
  large: 'large',
  medium: 'medium',
  small: 'small'
}

const Button = ({
  url,
  external,
  icon,
  iconSide = 'right',
  disabled,
  loading,
  hollow,
  destructive,
  fullWidth,
  submit,
  secondary,
  size = 'medium',
  textAlign = 'center',
  onClick,
  onKeyUp,
  onTouchEnter,
  children
}) => {
  let [state, setState] = useState({
    uiState: 'idle'
  })
  let theIcon = icon !== 'none' ? <Icon icon={icon} /> : null
  return (
    <button
      type={submit ? 'submit' : 'button'}
      className={`ds-button text-align-${textAlign} ${
        fullWidth ? 'display-block' : 'display-inline-flex'
      }`}
      disabled={disabled}
      ds-hollow={hollow ? 'true' : null}
      ds-primary={!secondary && !destructive && !disabled ? 'true' : null}
      ds-secondary={secondary && !destructive && !disabled ? 'true' : null}
      ds-icon={icon !== 'none' ? 'true' : null}
      ds-icon-side={icon !== 'none' ? iconSide : null}
      ds-size={size}
    >
      <span className="root">
        {iconSide === 'left' && icon !== 'none' && theIcon}
        <span>{children}</span>
        {iconSide === 'right' && icon !== 'none' && theIcon}
      </span>
    </button>
  )
}

export default Button

Button.defaultProps = {
  size: 'medium',
  textAlign: 'center',
  icon: 'none',
  iconSide: 'right'
}

Button.propTypes = {
  url: PropTypes.string,
  external: PropTypes.bool,

  icon: PropTypes.any,
  iconSide: PropTypes.oneOf(Object.keys(IconSides)),

  secondary: PropTypes.bool,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  hollow: PropTypes.bool,
  destructive: PropTypes.bool,
  fullWidth: PropTypes.bool,
  submit: PropTypes.bool,

  size: PropTypes.oneOf(Object.keys(Sizes)),
  textAlign: PropTypes.oneOf(Object.keys(Alignments)),

  onClick: PropTypes.func,
  onKeyUp: PropTypes.func,
  onTouchEnter: PropTypes.func
}
