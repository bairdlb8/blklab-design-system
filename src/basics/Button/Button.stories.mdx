import { Meta, Story, Props, Preview } from '@storybook/addon-docs/blocks'
import { withKnobs, select, boolean } from '@storybook/addon-knobs'
import { Compass } from 'react-feather'

import Button from './Button'

<Meta title="Basics / Button" component={Button} />

# Button

## Displays an image that represents a user or organization

Use an avatar for attributing actions or content to specific users. The user’s name should _always_ be present when using Button – either printed beside the avatar or in a tooltip.

export const Template = (args) => {
  return <Button {...args}>Click Me</Button>
}

<Preview withToolbar>
  <Story name="Primary" args={{}}>
    {Template.bind({})}
  </Story>
</Preview>

<Props of={Button} />

<Preview withToolbar>
  <Story name="Secondary" args={{ secondary: true }}>
    {Template.bind({})}
  </Story>
</Preview>
