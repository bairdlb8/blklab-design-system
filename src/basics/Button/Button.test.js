import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { toHaveNoViolations } from 'jest-axe'
import Button from './Button'

expect.extend(toHaveNoViolations)

describe('<Button />', () => {
  describe('url', () => {
    it('renders a link when present', () => {})
    it('renders a button when not', () => {})
  })

  describe('icon', () => {
    it('renders an icon when present', () => {})
  })

  describe('disabled', () => {
    it('is disabled when set to be', () => {})
  })

  describe('loading', () => {
    it('shows the loading indicator when set to loading', () => {})

    it('sets an alert role when loading', () => {})

    it('sets aria-busy on the button while loading', () => {})
  })

  describe('external', () => {
    it('has rel="noopener noreferrer" if is set', () => {})

    it('has target="_blank" when it is set', () => {})
  })

  describe('accessibility', () => {
    it('should not fail any accessibility tests', async () => {
      const { container } = render(<Button />)
      expect(await axe(container)).toHaveNoViolations()
    })

    it('sets an aria-label on the button', () => {})

    it('has aria-controls on the button', () => {})

    it('has aria-pressed on the button', () => {})

    it('sets an aria-label on the link if it has a url', () => {})
  })

  describe('onClick', () => {
    it('responds when onClick is called', () => {})
  })

  it('should take a snapshot', () => {
    const { asFragment } = render(<Button />)

    expect(asFragment(<Button />)).toMatchSnapshot()
  })
})
