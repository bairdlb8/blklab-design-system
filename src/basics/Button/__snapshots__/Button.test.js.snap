// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Button /> component should take a snapshot 1`] = `
<DocumentFragment>
  <button>
    Hi
  </button>
</DocumentFragment>
`;

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <button>
    Hi
  </button>
</DocumentFragment>
`;
